JOSE_paper/paper.md

```diff
@@ -0,0 +1,223 @@
+- 03: Improve code clarity with naming and source changes.
+- 04: Linting and static analysis of code.
+- 05: Writing docstrings and best use of comments.
+- 06: General techniques for better code (magic numbers, string formatting).
+- 00: The end point of the workshop - an improved version of the code in exercise 01.
+
+
+### Solutions
+
+Worked solutions to all of the exercises are provided in the form of completed
+notebooks including example output.
+These are available both in the repository and also as Colab instances.
+
+
+
+# Content Delivery
+
+The course has been designed to be flexible in terms of delivery, allowing
+it to be adapted to and reused in various setups.
+
+TODO Slides are interspersed with exercises.
+
+The main aspect we wish to emphasise in delivery is teaching in a _"code-along"_ fashion.
+This helps with engagement, participation, and understanding [@barba2022teaching] and
+is essential, we feel, to having a long-lasting benefit.
+This approach slows those leading the course towards the rate at which the participants
+are working, and illustrates through errors (whether intentional or not!) that even
+experienced coders are human and make mistakes.
+Such errors can illustrate common pitfalls and provide an opportunity to include
+the teaching of debugging approaches.
+More generally this approach helps emphasise RSE principles, as participants can
+see the live application of these ideas in practise.
+
+We also believe that there is sufficient guidance in the slides and repository
+documentation to follow the course alone, and we include a link to a recording of
+the workshop.
+This is, however, no substitute to in-person delivery where participants can ask
+questions, and successive workshops are continually improving.
+
+
+
+# Teaching experience
+
+One aspect of the material we have found useful is the ability to tailor the course to
+the workshop/audience.
+By breaking the slides into separate files by section and having standalone exercises
+for each section it is easy to add or remove material as desired.
+
+As discussed in the introduction, the course is written in Python for the broadest appeal,
+but is intended to educate researchers more generally on software engineering principles.
+We found it useful during delivery to reference tools and techniques applicable for
+other languages and systems as we went along.
+For example, where we introduce environment and dependency management in Python using
+virtual environments, we also mention spack as an analagous approach for those using HPC
+environments.
+We also discuss other language specific tools for formatting and linting/static analysis
+such as `clang-format` for C++.
+We feel that this is important to help participants see past the python specifics and
+understand the broader ideas and how they can be deployed in their own projects.
+
+TODO IDE and vimdiff
+
+TODO Could change the content to be domain specific in another area.
+
```

Comment:
Could talk about modularity, adapting the course content to the audience and the length of training session. 

Do we want to go into detail about how we do that in markdown with quarto, given that we mentioned using this? And then also that this is of course applicable in other environment as well, for example it works very similar with Latex?

Reply:
I added that briefly here